File: src/mlogix/logix/Expr.java
package mlogix.logix;

import mlogix.compiler.*;
import mlogix.compiler.struct.*;

import java.util.*;

//Expression
public abstract non-sealed class Expr extends ASTNode {
    protected Expr(Span span) {
        this.span = span;
    }

    public abstract Struct accept(SemanticAnalyzer.SemanticVisitor visitor);

    /* 字面量 */
    public static class Literal extends Expr {
        public final Token token;

        public Literal(Token token) {
            super(token.span);
            this.token = token;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }

    /* 标识符 */
    public static class Identifier extends Expr {
        public final Token token;

        public Identifier(Token token) {
            super(token.span);
            this.token = token;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }

    /* 一元运算 */
    public static class Unary extends Expr {
        public final Token operator;
        public final Expr expr;

        public Unary(Token operator, Expr expr) {
            super(Span.between(operator.span, expr.span));
            this.operator = operator;
            this.expr = expr;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }

    /* 二元运算 */
    public static class Binary extends Expr {
        public final Expr left;
        public final Token operator;
        public final Expr right;

        public Binary(Expr left, Token operator, Expr right) {
            super(Span.between(left.span, right.span));
            this.left = left;
            this.operator = operator;
            this.right = right;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }

    /* 数组 */
    public static class Array extends Expr {
        public final List<Expr> elements;

        public Array(Span span, List<Expr> elements) {
            super(span);
            this.elements = elements;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }

    /* 索引 */
    public static class Index extends Expr {
        public final Expr list;
        public final Expr index;

        public Index(Span span, Expr list, Expr index) {
            super(span);
            this.list = list;
            this.index = index;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }

    public static class Range extends Expr {
        public final Expr left;
        public final Token operator;
        public final Expr right;

        public Range(Span span, Expr left, Token operator, Expr right) {
            super(span);
            this.left = left;
            this.operator = operator;
            this.right = right;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }

    /* 函数调用 func(...) */
    public static class Call extends Expr {
        public final Expr callee;
        public final List<Expr> arguments;

        public Call(Span span, Expr callee, List<Expr> arguments) {
            super(span);
            this.callee = callee;
            this.arguments = arguments;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }

    /* 获取字段 struct.field  struct.func */
    public static class Get extends Expr {
        public final Expr object;
        public final Expr field;

        public Get(Expr object, Expr field) {
            super(Span.between(object.span, field.span));
            this.object = object;
            this.field = field;
        }

        @Override
        public Struct accept(SemanticAnalyzer.SemanticVisitor visitor) {
            return visitor.visit(this);
        }
    }
}

File: src/mlogix/mlogix/ast/Expr.kt
package mlogix.mlogix.ast

import arc.struct.Seq
import mlogix.mlogix.token.Token
import mlogix.span.Span

//Expression
abstract class Expr(span: Span) : ASTNode(span) {
    /**
     * 字面量
     */
    data class Literal(val token: Token) : Expr(token.span)

    /**
     * 标识符
     */
    data class Identifier(val token: Token) : Expr(token.span)

    /**
     * 元组
     */
    data class Tuple(override val span: Span, val elements: Seq<Expr>) : Expr(span)

    /**
     * 需要保证annotations.size > 0
     */
    data class Annotation(
        val expr: Expr,
        val annotations: Seq<Expr>
    ) : Expr(Span.between(expr, annotations[annotations.size - 1]))

    /**
     * 一元运算
     */
    data class Unary(val operator: Token, val expr: Expr) : Expr(Span.between(operator.span, expr.span))

    /**
     * 二元运算
     */
    data class Binary(val left: Expr, val operator: Token, val right: Expr) : Expr(
        Span.between(
            left.span, right.span
        )
    )

    /**
     * 数组
     */
    data class Array(override val span: Span, val elements: Seq<Expr>) : Expr(span)

    /**
     * 索引
     */
    data class Index(override val span: Span, val list: Expr, val index: Expr) : Expr(span)


    /**
     * 范围
     */
    data class Range(override val span: Span, val left: Expr?, val operator: Token, val right: Expr?) : Expr(span)

    /**
     * 函数调用 func(...)
     */
    data class Call(override val span: Span, val callee: Expr, val arguments: Seq<Expr>) : Expr(span)

    /**
     * 获取字段 type.field  type.func
     */
    data class Get(val obj: Expr, val field: Expr) : Expr(Span.between(obj.span, field.span))

    /**
     * 错误恢复占位符
     */
    data class ErrorExpr(override val span: Span) : Expr(span)
}